vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -30,83 +34,273 @@ type internal FSharpNavigableItem(document: Document, textSpan: TextSpan) =
-            // REVIEW: 
-            let fileName = try System.IO.Path.GetFullPath(range.FileName) with _ -> range.FileName
-            let refDocumentIds = document.Project.Solution.GetDocumentIdsWithFilePath(fileName)
-            if not refDocumentIds.IsEmpty then 
-                let refDocumentId = refDocumentIds.First()
-                let refDocument = document.Project.Solution.GetDocument(refDocumentId)
-                let! refSourceText = refDocument.GetTextAsync(cancellationToken)
-                let refTextSpan = CommonRoslynHelpers.FSharpRangeToTextSpan(refSourceText, range)
-                results.Add(FSharpNavigableItem(refDocument, refTextSpan))
-            return results.AsEnumerable()
-         }
-         |> Async.map (Option.defaultValue Seq.empty)
-         |> CommonRoslynHelpers.StartAsyncAsTask cancellationToken
+            let! projectOptions = projectInfoManager.TryGetOptionsForEditingDocumentOrProject originDocument
+            let defines = CompilerEnvironment.GetCompilationDefinesForEditing (originDocument.FilePath, projectOptions.OtherOptions |> Seq.toList)
 
-    interface IGoToDefinitionService with
-        member this.FindDefinitionsAsync(document: Document, position: int, cancellationToken: CancellationToken) =
-            this.FindDefinitionsAsyncAux(document, position, cancellationToken)
+            let originTextSpan = CommonRoslynHelpers.FSharpRangeToTextSpan (sourceText, range)
+            let position = originTextSpan.Start
+
+            let! lexerSymbol = 
+                CommonHelpers.getSymbolAtPosition 
+                    (originDocument.Id, sourceText, position, originDocument.FilePath, defines, SymbolLookupKind.Greedy)
+            
+            let textLinePos = sourceText.Lines.GetLinePosition position
+            let fcsTextLineNumber = Line.fromZ textLinePos.Line
+            let lineText = (sourceText.Lines.GetLineFromPosition position).ToString()  
+            
+            let! _, _, checkFileResults = 
+                checker.ParseAndCheckDocument (originDocument,projectOptions,allowStaleResults=true,sourceText=sourceText)
+            let idRange = lexerSymbol.Ident.idRange
 
-        member this.TryGoToDefinition(document: Document, position: int, cancellationToken: CancellationToken) =
-            let definitionTask = this.FindDefinitionsAsyncAux(document, position, cancellationToken)
+            let! fsSymbolUse = checkFileResults.GetSymbolUseAtLocation (fcsTextLineNumber, idRange.EndColumn, lineText, lexerSymbol.FullIsland)
+            let symbol = fsSymbolUse.Symbol
             
-            // REVIEW: document this use of a blocking wait on the cancellation token, explaining why it is ok
-            definitionTask.Wait(cancellationToken)
+            // if the tooltip was spawned in an implementation file and we have a range targeting
```

Comment:
Maybe it would be useful to have separate function(s) returning both .fsi and .fs locations for a symbol?